### PR TITLE
dasharo-security/tpm-support.robot: Test TPM status in BIOS

### PR DESCRIPTION
**Pull request by m-iwanicki**

### CODE REVIEW

**Comment by PLangowski** [comment]
The test works properly and passes

```
==============================================================================
TPM001.004 TPM Support (BIOS) :: This test aims to verify that the... | PASS |
------------------------------------------------------------------------------
```

File: lib/tpm-support-lib.robot [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+Enter TCG2 Menu
+    [Documentation]    This keyword enters TCG2 menu after the platform was powered on.
+    ${setup_menu}=    Enter Setup Menu Tianocore And Return Construction
+    ${device_mgr_menu}=    Enter Submenu From Snapshot And Return Construction
+    ...    ${setup_menu}
+    ...    Device Manager
+    Enter Submenu From Snapshot    ${device_mgr_menu}    TCG2 Configuration
+
+Enter TCG2 Menu And Return Construction
+    [Documentation]    This keyword enters TCG2 menu after the platform was powered on. Returns TCG2 menu construction.
```
Comment (TomaszAIR) [inline lib/tpm-support-lib.robot:46]: @m-iwanicki please wrap at 80 signs

File: lib/tpm-support-lib.robot [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    # Go back to top
+    FOR    ${i}    IN RANGE    3
+        Sleep    0.1
+        Press Key N Times    1    ${PAGEUP}
+    END
+    ${menu_dict}=    Convert List Of Pairs Into Dictionary    ${menu}
+    RETURN    ${menu_dict}
+
+Enter TCG2 Menu
+    [Documentation]    This keyword enters TCG2 menu after the platform was powered on.
```
Comment (TomaszAIR) [inline lib/tpm-support-lib.robot:38]: @m-iwanicki please wrap at 80 signs

File: lib/tpm-support-lib.robot [hunk 3]
```diff
@@ -0,0 +1,58 @@
+*** Settings ***
+Documentation       Collection of keywords related to UEFI TPM support
+
+Library             ./tpm-support-lib.py
+Resource            ../keywords.robot
+Resource            ../keys.robot
+
+
+*** Keywords ***
+# TODO: Most options are split in 2 lines either because name is too long or
```
Comment (TomaszAIR) [inline lib/tpm-support-lib.robot:10]: @m-iwanicki what we should do with this TODO?

File: lib/tpm-support-lib.robot [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#    they have too many arguments. Get TCG2 Menu Construction currently
+#    returns those as different list items
```
Comment (TomaszAIR) [inline lib/tpm-support-lib.robot:12]: @m-iwanicki I think there could be more submenus where multilines happenes, do we have some idea how to use these keywords in other then TCG2 menus?

**Review by TomaszAIR — CHANGES_REQUESTED** [review]
@m-iwanicki I left review, could you also check if this branch is correctly rebased on top of `encrypted-rootfs-release` branch?

File: lib/tpm-support-lib.robot (hunk 3 above)
Comment (m-iwanicki) [reply lib/tpm-support-lib.robot:10]: I forgot to delete it, currently keyword returns dictionary containing 3 pages of merged options/values

File: lib/tpm-support-lib.robot (hunk 4 above)
Comment (m-iwanicki) [reply lib/tpm-support-lib.robot:12]: @TomaszAIR Keyword should probably work for many other menus with minimal changes.
@{additional_remove} should be argument or contain all possible values to be removed for every menu (there are not that many).
Number of pages could be determined by presence of down arrow

File: lib/tpm-support-lib.robot (hunk 3 above)
Comment (m-iwanicki) [reply lib/tpm-support-lib.robot]: @TomaszAIR fixed in [ca43cb6](https://github.com/Dasharo/open-source-firmware-validation/pull/164/commits/ca43cb6284650227c50aca4e3e4cc2c83598b05c)

File: lib/tpm-support-lib.robot (hunk 2 above)
Comment (m-iwanicki) [reply lib/tpm-support-lib.robot]: fixed in [ca43cb6](https://github.com/Dasharo/open-source-firmware-validation/pull/164/commits/ca43cb6284650227c50aca4e3e4cc2c83598b05c)

**Comment by m-iwanicki** [comment]
> @m-iwanicki I left review, could you also check if this branch is correctly rebased on top of `encrypted-rootfs-release` branch?

# git rebase origin/encrypted-rootfs-release 
Current branch tpm-firmware-test is up to date.

File: lib/tpm-support-lib.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+from robot.api.deco import keyword
+
+
+@keyword("Parse TCG2 Menu Snapshot Into Construction")
+def parse_tcg2_menu_snapshot_into_construction(
+    menu: str, remove_lines=None, merge_empty_options=True
+):
+    """Breaks grabbed menu data into list of (option, value) pairs.
+
+    Correctly merges multiline names and values if values end in comma
```
Comment (TomaszAIR) [inline lib/tpm-support-lib.py:12]: @m-iwanicki I am not sure if I understand this description, this part about comma is only for `values` part of menu?

Comment (m-iwanicki) [reply lib/tpm-support-lib.py:12]: If value ends with comma that means that next line will be continuation of this line, i.e.

| | |
|-|-|
| Bios Supported Hash | SHA1, SHA256, SHA384, |
|Algorithm | SHA512 |

Thanks to that comma it's possible to merge those 2 lines into one part.

**Comment by TomaszAIR** [comment]
Test result

```
λ robot -L TRACE -v rte_ip:127.0.0.1 -v snipeit:no -v config:qemu -t "TPM001.004*" dasharo-security/tpm-support.robot                                  
==============================================================================
Tpm-Support                                                                   
==============================================================================
TPM001.004 TPM Support (BIOS) :: This test aims to verify that the... | PASS |
------------------------------------------------------------------------------
Tpm-Support                                                           | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  /home/tzyjewski/files/open-source-firmware-validation/output.xml
Log:     /home/tzyjewski/files/open-source-firmware-validation/log.html
Report:  /home/tzyjewski/files/open-source-firmware-validation/report.html
```

Merging

File: lib/tpm-support-lib.robot (hunk 4 above)
Comment (TomaszAIR) [reply lib/tpm-support-lib.robot]: And this is something worth to left here as TODO. Added here https://github.com/Dasharo/open-source-firmware-validation/pull/164/commits/7dbf68a4e47fbe29973912d42dc78c4d6aa31249

**Pull request merged**